lang/src/org/partiql/lang/eval/physical/operators/Accumulator.kt

```diff
@@ -0,0 +1,184 @@
+
+    override fun nextValue(value: ExprValue) {
+        checkIsNumberType(funcName = "SUM", value = value)
+        if (sum == null) sum = 0.0
+        this.sum = value.numberValue() + this.sum!!
+    }
+
+    override fun compute(): ExprValue {
+        val totalSum = sum?.let { bigDecimalOf(it) }
+        return totalSum?.exprValue(valueFactory) ?: valueFactory.nullValue
+    }
+}
+
+internal class AccumulatorAvg(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var sum: Number = 0.0
+    var count: Long = 0L
+
+    override fun nextValue(value: ExprValue) {
+        checkIsNumberType(funcName = "AVG", value = value)
+        this.sum += value.numberValue()
+        this.count += 1L
+    }
+
+    override fun compute(): ExprValue = when (count) {
+        0L -> valueFactory.nullValue
+        else -> (sum / bigDecimalOf(count)).exprValue(valueFactory)
+    }
+}
+
+internal class AccumulatorMax(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var max: ExprValue = valueFactory.nullValue
+
+    override fun nextValue(value: ExprValue) {
+        max = comparisonAccumulator(NaturalExprValueComparators.NULLS_LAST_DESC)(max, value)
+    }
+
+    override fun compute(): ExprValue = max
+}
+
+internal class AccumulatorMin(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var min: ExprValue = valueFactory.nullValue
+
+    override fun nextValue(value: ExprValue) {
+        min = comparisonAccumulator(NaturalExprValueComparators.NULLS_LAST_ASC)(min, value)
+    }
+
+    override fun compute(): ExprValue = min
+}
+
+internal class AccumulatorCount(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    var count: Long = 0L
+
+    override fun nextValue(value: ExprValue) {
+        this.count += 1L
+    }
+
+    override fun compute(): ExprValue = count.exprValue(valueFactory)
+}
+
+internal class AccumulatorGroupAs(
+    internal val valueFactory: ExprValueFactory,
+    internal override val filter: (ExprValue) -> Boolean
+) : Accumulator(filter = filter) {
+
+    val exprValues = mutableListOf<ExprValue>()
+
+    override fun nextValue(value: ExprValue) {
+        exprValues.add(value)
+    }
+
+    override fun compute(): ExprValue = valueFactory.newBag(exprValues)
+}
+
+private fun comparisonAccumulator(comparator: NaturalExprValueComparators): (ExprValue?, ExprValue) -> ExprValue =
+    { left, right ->
+        when {
+            left == null || comparator.compare(left, right) > 0 -> right
+            else -> left
+        }
+    }
+
+internal fun checkIsNumberType(funcName: String, value: ExprValue) {
+    if (!value.type.isNumber) {
```

Comment:
Should `NULL` or `MISSING` be allowed here as well?

Reply:
No, see `Accumulator.next()`. We essentially filter out `NULL`, `MISSING`, and non-distinct values if requested.

Reply:
I see. Sorry I missed that.